heartbeat/mariadb-semi-sync

```diff
@@ -0,0 +1,1064 @@
+    for node in $OCF_RESKEY_node_list; do
+
+        local node_gtid=$(read_gtid $node)
+        if [ -z "$node_gtid" ]; then
+            continue
+        fi
+
+        # Got a valid gtid, increment node count
+        node_count=$(($node_count+1))
+
+        # Check if this is a good master candidate
+        if greater_than_gtid $node_gtid $highest_gtid; then
+            master_candidate=$node
+            highest_gtid=$node_gtid
+        fi
+    done
+
+    # If we managed to query a sufficient number of nodes
+    # then set a master
+    if [ $node_count -ge $expected_node_count ]; then
+        ocf_log info "Promoting $master_candidate to master, highest gtid $highest_gtid, queried $node_count nodes."
+        $CRM_MASTER -v 100 -N $master_candidate
+    else
+        ocf_log info "Not enough nodes ($node_count) contributed to select a master, need $expected_node_count nodes."
+    fi
+}
+
+set_read_only() {
+    # Sets or unsets read-only mode. Accepts one boolean as its
+    # optional argument. If invoked without any arguments, defaults to
+    # enabling read only mode. Should only be set in master/slave
+    # setups.
+    # Returns $OCF_SUCCESS if the operation succeeds, or
+    # $OCF_ERR_GENERIC if it fails.
+    local ro_val
+    if ocf_is_true $1; then
+        ro_val="on"
+    else
+        ro_val="off"
+    fi
+    ocf_run $MYSQL $MYSQL_OPTIONS_REPL \
+        -e "SET GLOBAL read_only=${ro_val}"
+}
+
+get_read_only() {
+    # Check if read-only is set
+    local read_only_state
+
+    read_only_state=$($MYSQL $MYSQL_OPTIONS_REPL \
+        -e "SHOW VARIABLES" | grep -w read_only | awk '{print $2}')
+
+    if [ "$read_only_state" = "ON" ]; then
+        return 0
+    else
+        return 1
+    fi
+}
+
+is_slave() {
+    # Determine whether the machine is currently running as a MariaDB
+    # slave, as determined per SHOW SLAVE STATUS. Returns 1 if SHOW
+    # SLAVE STATUS creates an empty result set, 0 otherwise.
+    local rc
+
+    # Check whether this machine should be slave
+    if ! get_read_only; then
+        return 1
+    fi
+   
+    get_slave_info
+    rc=$?
+    rm -f $tmpfile
+
+    if [ $rc -eq 0 ]; then
+       # show slave status is not empty
+       # Is the slave sql thread running, then we are a slave!
+       if [ "$slave_sql" == 'Yes' ]; then
+          return 0
+       else
+          return 1
+       fi
+    else
+       # "SHOW SLAVE STATUS" returns an empty set if instance is not a
+       # replication slave
+       return 1
+    fi
+}
+
+parse_slave_info() {
+    # Extracts field $1 from result of "SHOW SLAVE STATUS\G" from file $2
+    sed -ne "s/^.* $1: \(.*\)$/\1/p" < $2
+}
+
+get_slave_info() {
+    # Warning: this sets $tmpfile and LEAVES this file! You must delete it after use!
+    local mysql_options
+            
+    if [ "$master_log_file" -a "$master_host" ]; then
+        # variables are already defined, get_slave_info has been run before
+        return $OCF_SUCCESS
```

Comment:
tmpfile isn't set in this case. maybe it should explicitly set tmpfile="" and then the remover can check -n $tmpfile.